From: olh@suse.de
Subject: [PATCH] Fix reference to deallocated memory in sd.c

# ChangeSet
#   2005/01/17 12:03:49-06:00 stern@rowland.harvard.edu 
#   [PATCH] Fix reference to deallocated memory in sd.c
#   
#   This patch:
#   
#   http://linux-scsi.bkbits.net:8080/scsi-for-linus-2.6/cset@1.2034.95.5?nav=index.html|src/|src/drivers|src/drivers/scsi|related/drivers/scsi/sd.c
#   
#   is causing almost as much trouble as it fixed.  If kref_put() drops the
#   last reference to the scsi_disk (this happens when the device file is
#   closed after the device has been hot-unplugged) then the call to
#   scsi_device_put() will take its argument from an area of memory that has
#   been deallocated.
#   
#   Signed-off-by: Alan Stern <stern@rowland.harvard.edu>
#   Signed-off-by: James Bottomley <James.Bottomley@SteelEye.com>
# 
# drivers/scsi/sd.c
#   2005/01/14 10:01:14-06:00 stern@rowland.harvard.edu +3 -1
#   Fix reference to deallocated memory in sd.c
diff -Nru a/drivers/scsi/sd.c b/drivers/scsi/sd.c
--- a/drivers/scsi/sd.c	2005-01-23 22:28:26 -08:00
+++ b/drivers/scsi/sd.c	2005-01-23 22:28:26 -08:00
@@ -197,9 +198,11 @@
 
 static void scsi_disk_put(struct scsi_disk *sdkp)
 {
+	struct scsi_device *sdev = sdkp->device;
+
 	down(&sd_ref_sem);
 	kref_put(&sdkp->kref, scsi_disk_release);
-	scsi_device_put(sdkp->device);
+	scsi_device_put(sdev);
 	up(&sd_ref_sem);
 }
 
